fabric/src/test/java/TestRegexs.java

```diff
@@ -0,0 +1,133 @@
+        assertMatchesAny(pattern, "§a✔ §7Unlocked Abilities: §f14§7/15")
+                .assertGroupIs(1, "14")
+                .assertGroupIs(2, "15");
+        assertMatchesAny(pattern, "§a✔ §7Unlocked Abilities: §f2§7/16")
+                .assertGroupIs(1, "2")
+                .assertGroupIs(2, "16");
+        assertMatchesAny(pattern, "§a✔ §7Unlocked Abilities: §f0§7/15")
+                .assertGroupIs(1, "0")
+                .assertGroupIs(2, "15");
+    }
+
+    @Test
+    public void testPricePattern() {
+        Pattern pattern = TestPrivateFieldsUtil.getPrivateStaticRegexPattern(BulkBuyFeature.class, "PRICE_PATTERN");
+        assertMatchesAny(pattern, "§6 - §a✔ §f10§7²");
+        assertMatchesAny(pattern, "§6 - §a✔ §f1483§7²");
+        assertMatchesAny(pattern, "§6 - §c✖ §f244§7²");
+        assertMatchesAny(pattern, "§6 - §c✖ §f1§7²");
+        assertMatchesAny(pattern, "§6 - §a✔ §f24§7²");
+    }
+
+    @Test
+    public void testGuildMessagePattern() {
+        Pattern pattern = TestPrivateFieldsUtil.getPrivateStaticRegexPattern(
+                GuildRankReplacementFeature.class, "GUILD_MESSAGE_PATTERN");
+        assertMatchesAny(pattern, "§3[§b★★★★★§3§oDisco reroller§3]");
+        assertMatchesAny(pattern, "§3[§b★★★★★§3§oafKing§r§3]§");
+        assertMatchesAny(pattern, "§3[§b★★★★§3§obol§r§3] ");
+    }
+
+    @Test
+    public void testRecruitUsernamePattern() {
+        Pattern pattern = TestPrivateFieldsUtil.getPrivateStaticRegexPattern(
+                GuildRankReplacementFeature.class, "RECRUIT_USERNAME_PATTERN");
+        assertMatchesAny(pattern, "§3[_user0name_").assertGroupIs(1, "_user0name_");
+        assertMatchesNone(pattern, "§3[WAR]");
+        assertMatchesAny(pattern, "§3[ummmmmmmmm]").assertGroupIs(1, "ummmmmmmmm");
+    }
+
+    public record MatcherAssertions(Matcher matcher) {
+        public MatcherAssertions assertGroupIs(int group, String expected) {
+            String resultGroup = matcher.group(group);
+            Assertions.assertEquals(
+                    expected,
+                    resultGroup,
+                    String.format(
+                            "asserted regex group should be equal to string,\ngroup:\t\t%d\nexpected:\t%s\nresult:\t\t%s%n",
+                            group, expected, resultGroup));
+            return this;
+        }
+
+        public MatcherAssertions assertGroupIs(String group, String expected) {
+            String resultGroup = matcher.group(group);
+            Assertions.assertEquals(
+                    expected,
+                    resultGroup,
+                    String.format(
+                            "asserted regex group should be equal to string,\ngroup:\t\t%s\nexpected:\t%s\nresult:\t\t%s%n",
+                            group, expected, resultGroup));
+            return this;
+        }
+    }
+
+    public MatcherAssertions assertMatchesAny(Pattern pattern, String string) {
```

Comment:
I'd like to break out all helper methods to a separate "test lib" class, and then have the actual tests structured in a way that reasonably mirrors the code base. That it, it should be an easy to figure out, 1-to-1 mapping, on how to go to the test from the source code and vice versa.

Reply:
I'm not sure what is the most modern and common junit preference for doing this, if we should have like a utility class with static members, or like an abstract superclass with protected helpers, and then subclass it for the actual tests. When I last used junit seriously, the latter was the mode du jour in junit coding.